common4j/src/main/com/microsoft/identity/common/java/storage/StringSeparatedMultiTypeNameValueStorage.java

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.storage;
+
+import com.microsoft.identity.common.java.cache.IMultiTypeNameValueStorage;
+import com.microsoft.identity.common.java.crypto.IKeyAccessor;
+import com.microsoft.identity.common.java.interfaces.IPerSeparatorMultiTypeNameValueStore;
+import com.microsoft.identity.common.java.interfaces.IPlatformComponents;
+import com.microsoft.identity.common.java.util.ported.Predicate;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.NonNull;
+
+/**
+ * A platform agnostic implementation of {@link IPerSeparatorMultiTypeNameValueStore} where the
+ * separator is of type {@link String}. The internal storage mechanism here could be backed by
+ * anything as determined by the {@link IPlatformComponents#getFileStore(String)} or even be an
+ * encrypted variant as determined by {@link IPlatformComponents#getEncryptedFileStore(String, IKeyAccessor)}.
+ */
+@AllArgsConstructor
+public class StringSeparatedMultiTypeNameValueStorage implements IPerSeparatorMultiTypeNameValueStore<String> {
```

Comment:
Wondering if we can turn this into a utility class by making getStoreForSeparator() abstract.

by doing that, we should not need the interface.

Reply:
IMO, most classes should be backed by an interface and with this one I think it is needed to allow someone to take it and implement in a completely different way.

I did go ahead and added an abstract implementation as you suggested but did so on top of the interface. Addressed here: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1810/commits/2830fc03360f3cadff042300f228c5d0ca03ca15